interfaces/apparmor/backend.go

```diff
@@ -285,6 +286,31 @@ func profileGlobs(snapName string) []string {
 	return []string{interfaces.SecurityTagGlob(snapName), nsProfile(snapName)}
 }
 
+// Determine if a profile filename is removable during core refresh/rollback.
+// This is needed because core devices are also special, the apparmor cache
+// gets confused too easy, especially at rollbacks, so we delete the cache. See
+// Setup(), below. Some systems employ a unified cache directory where all
+// apparmor cache files are stored under one location so ensure we don't remove
+// the snap profiles since snapd manages them elsewhere and instead only remove
+// snap-confine and system profiles (eg, as shipped by distro package manager
+// or created by the administrator). snap-confine profiles are like the
+// following:
+// - usr.lib.snapd.snap-confine.real
+// - usr.lib.snapd.snap-confine
+// - snap.core.NNNN.usr.lib.snapd.snap-confine
+// - var.lib.snapd.snap.core.NNNN.usr.lib.snapd.snap-confine
+// - snap-confine.core.NNNN
+// TODO: also the "snapd" snap here soon
+func profileIsRemovableOnCoreSetup(fn string) bool {
+	bn := path.Base(fn)
+	if strings.HasPrefix(bn, ".") {
+		return false
+	} else if strings.HasPrefix(bn, "snap") && !strings.HasPrefix(bn, "snap-confine.core.") && !strings.Contains(bn, "usr.lib.snapd.snap-confine") {
```

Comment:
Should we do `usr.libexec.snapd` as well?

Reply:
This will only run on core devices (not classic) so we should never have libexec.snapd here (even with a fedora base snapd will still come from the snapd snap with the "usr/lib" layout).

Reply:
+1